File: PythonCode/lintcode/easy/WordPattern.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# @Time    : 2018/3/13 14:19
# @Author  : bearyang
# @Site    : 
# @File    : WordPattern.py
# @Software: PyCharm
# Lintcode 828. Word Pattern
# http://www.lintcode.com/zh-cn/problem/word-pattern/
# Leetcode 290. Word Pattern
# https://leetcode.com/problems/word-pattern/description/


class Solution:
    """
    @param pattern: a string, denote pattern string
    @param str: a string, denote matching string
    @return: an boolean, denote whether the pattern string and the matching string match or not
    """
    def wordPattern(self, pattern, str):
        strArray = str.split(' ')
        if len(pattern) != len(strArray):
            return False
        length = len(pattern)
        map = {}
        for i in range(length):
            ch = pattern[i]
            if ch in map.keys():
                if map.get(ch) != strArray[i]:
                    return False
            else:
                if strArray[i] in map.values():
                    return False
                map[ch] = strArray[i]
        return True


test = Solution()
pattern1 = 'abba'
str1 = 'dog cat cat dog'
print(test.wordPattern(pattern1, str1))

pattern2 = 'abba'
str2 = 'dog dog dog dog'
print(test.wordPattern(pattern2, str2))

File: JsCode/lintcode/easy/WordPattern.js
/**
 * Lintcode 828. Word Pattern
 * http://www.lintcode.com/zh-cn/problem/word-pattern/
 * Leetcode 290. Word Pattern
 * https://leetcode.com/problems/word-pattern/description/
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function(pattern, str) {
    var strArray = str.split(' ');
    if (pattern.length !== strArray.length) {
        return false;
    }
    var length = pattern.length;
    var map = {};
    for(var i = 0; i < length; i++) {
        var ch = pattern[i];
        if (ch in map) {
            if (map[ch] !== strArray[i]) {
                return false;
            }
        } else {
            for (var key in map) {
                if (strArray[i] === map[key]) {
                    return false;
                }
            }
            map[ch] = strArray[i];
        }
    }
    return true;
};


var pattern1 = 'abba';
var str1 = 'dog cat cat dog';
console.log(wordPattern(pattern1, str1));

var pattern2 = 'abba';
var str2 = 'dog dog dog dog';
console.log(wordPattern(pattern2, str2));